python-frontend/src/main/java/org/sonar/python/semantic/v2/TypeInferenceV2.java

```diff
@@ -279,15 +277,20 @@ public void visitQualifiedExpression(QualifiedExpression qualifiedExpression) {
 
   @Override
   public void visitName(Name name) {
-    SymbolV2 symbolV2 = name.symbolV2();
-    if (symbolV2 == null) {
-      return;
-    }
-    List<PythonType> types = new ArrayList<>();
-    for (UsageV2 usage : symbolV2.usages()) {
-      if (usage.kind().equals(UsageV2.Kind.GLOBAL_DECLARATION)) {
+    Optional.of(name)
+      .map(Name::symbolV2)
+      .map(SymbolV2::usages)
+      .map(TypeInferenceV2::getUsagesType)
+      .or(() -> projectLevelTypeTable.getModule().resolveMember(name.name()))
```

Comment:
I think this is a subtle yet impactful change from the previous implementation.

In the previous implementation, any variable that either has multiple assignments or a global declaration usage would not be assigned a type.

Now we fall back to `builtins` instead, which means that in the following (admittedly convoluted) code:

```
global int
int = 42
int = "hello"
int
```

The final `int` would be considered to be `builtins.int` despite the multiple reassignments.

I think the fallback to builtins should only happen if a symbol has no binding usages.

I'm a bit unhappy about the focus of our discussion regarding the imperative loop vs the optional chain. I think the main focus should be on functional behavior rather than technical implementation and I believe this case was an example where we can improve.

Taking the previous implementation as a reference, I believe that a simple

```
if (types.size().isEmpty() {
  // the underlying assumption is that each binding usage will have a type and add to the list
  // if it's not the case, a simple boolean "hasNoBindingUsage" computed during the iteration would do the trick
  setTypeToName(name, projectLevelTypeTable.getModule().resolveMember(name.name()));
}
```
at the end of the method would have achieved what you're trying to do, without requiring an additional iteration over `usages` and without risking an undesired functional change.

I also think we are missing tests for this (like variations of the convoluted case I showed), which is hidden due to the optional chain.

Again, I have nothing against optional chains, but they do have limitations that need to be acknowledged and I don't think we should aim to refactor everything to these just for the sake of it.